Build fails with clang++ 10/Apple g++

Build fails with latest clang (Apple LLVM version 10.0.1 (clang-1001.0.46.4)) and g++ from Apple.  With MacOS Mojave 10.14.6. 
Build with g++ 9.1.0 (built on system through Homebrew) succeeds.

Short version: a lot of errors that look like:
 
`In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:721:71: error: invalid operands to binary expression ('const
      gko::detail::IteratorFactory<int, float>::element' and 'const gko::detail::IteratorFactory<int, float>::element')
    bool operator()(const _T1& __x, const _T1& __y) const {return __x < __y;}` 
and 
`In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3718:20: error: no matching function for call to '__sort3'
    unsigned __r = __sort3<_Compare>(__x1, __x2, __x3, __c);`

<details><summary>
Click for long version (full output): </summary>

[  0%] Built target ginkgo_omp_device
[  0%] Built target ginkgo_cuda_device
[  0%] Built target ginkgo_reference_device
[  0%] Built target make_run_all_benchmarks
[  1%] Built target ginkgo_cuda
[  1%] Building CXX object reference/CMakeFiles/ginkgo_reference.dir/matrix/csr_kernels.cpp.o
[  2%] Built target ginkgo_omp
[  2%] Built target generate_ginkgo_header
In file included from /Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:33:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/core/matrix/csr_kernels.hpp:37:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/include/ginkgo/core/base/types.hpp:43:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/complex:247:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/sstream:174:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ostream:138:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:500:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:176:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:721:71: error: invalid operands to binary expression ('const
      gko::detail::IteratorFactory<int, float>::element' and 'const gko::detail::IteratorFactory<int, float>::element')
    bool operator()(const _T1& __x, const _T1& __y) const {return __x < __y;}
                                                                  ~~~ ^ ~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3931:17: note: in instantiation of member function
      'std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element>::operator()'
      requested here
            if (__comp(*--__last, *__first))
                ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &,
      gko::detail::IteratorFactory<int, float>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, float>::element,
      gko::detail::IteratorFactory<int, float>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/utility:594:1: note: candidate template ignored: could not match
      'pair<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator< (const pair<_T1,_T2>& __x, const pair<_T1,_T2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:702:1: note: candidate template ignored: could not match
      'reverse_iterator<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const reverse_iterator<_Iter1>& __x, const reverse_iterator<_Iter2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1143:1: note: candidate template ignored: could not match
      'move_iterator<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const move_iterator<_Iter1>& __x, const move_iterator<_Iter2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1515:1: note: candidate template ignored: could not match
      '__wrap_iter<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const __wrap_iter<_Iter1>& __x, const __wrap_iter<_Iter2>& __y) _NOEXCEPT_DEBUG
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/tuple:1182:1: note: candidate template ignored: could not match
      'tuple<type-parameter-0-0...>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const tuple<_Tp...>& __x, const tuple<_Up...>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2970:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator< (const unique_ptr<_T1, _D1>& __x, const unique_ptr<_T2, _D2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3028:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const unique_ptr<_T1, _D1>& __x, nullptr_t)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3037:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(nullptr_t, const unique_ptr<_T1, _D1>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4808:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const shared_ptr<_Tp>& __x, const shared_ptr<_Up>& __y) _NOEXCEPT
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4878:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(const shared_ptr<_Tp>& __x, nullptr_t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4886:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, float>::element'
operator<(nullptr_t, const shared_ptr<_Tp>& __x) _NOEXCEPT
^
In file included from /Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:33:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/core/matrix/csr_kernels.hpp:37:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/include/ginkgo/core/base/types.hpp:43:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/complex:247:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/sstream:174:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ostream:138:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:500:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:176:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3718:20: error: no matching function for call to '__sort3'
    unsigned __r = __sort3<_Compare>(__x1, __x2, __x3, __c);
                   ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3938:20: note: in instantiation of function template specialization
      'std::__1::__sort4<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element>
      &, gko::detail::IteratorFactory<int, float>::Iterator>' requested here
            _VSTD::__sort4<_Compare>(__first, __first+1, __first+2, --__last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &,
      gko::detail::IteratorFactory<int, float>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, float>::element,
      gko::detail::IteratorFactory<int, float>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3744:20: error: no matching function for call to '__sort4'
    unsigned __r = __sort4<_Compare>(__x1, __x2, __x3, __x4, __c);
                   ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3941:20: note: in instantiation of function template specialization
      'std::__1::__sort5<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element>
      &, gko::detail::IteratorFactory<int, float>::Iterator>' requested here
            _VSTD::__sort5<_Compare>(__first, __first+1, __first+2, __first+3, --__last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &,
      gko::detail::IteratorFactory<int, float>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, float>::element,
      gko::detail::IteratorFactory<int, float>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3715:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort4(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3809:5: error: no matching function for call to '__sort3'
    __sort3<_Compare>(__first, __first+1, __j, __comp);
    ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3946:20: note: in instantiation of function template specialization
      'std::__1::__insertion_sort_3<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int,
      float>::element> &, gko::detail::IteratorFactory<int, float>::Iterator>' requested here
            _VSTD::__insertion_sort_3<_Compare>(__first, __last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &,
      gko::detail::IteratorFactory<int, float>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, float>::element,
      gko::detail::IteratorFactory<int, float>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3842:9: error: no matching function for call to '__sort3'
        _VSTD::__sort3<_Compare>(__first, __first+1, --__last, __comp);
        ^~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__config:508:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4072:32: note: in instantiation of function template specialization
      'std::__1::__insertion_sort_incomplete<std::__1::__less<gko::detail::IteratorFactory<int, float>::element,
      gko::detail::IteratorFactory<int, float>::element> &, gko::detail::IteratorFactory<int, float>::Iterator>' requested here
            bool __fs = _VSTD::__insertion_sort_incomplete<_Compare>(__first, __i, __comp);
                               ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &,
      gko::detail::IteratorFactory<int, float>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, float>::element,
      gko::detail::IteratorFactory<int, float>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, float>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3845:9: error: no matching function for call to '__sort4'
        _VSTD::__sort4<_Compare>(__first, __first+1, __first+2, --__last, __comp);
        ^~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__config:508:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3715:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort4(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3848:9: error: no matching function for call to '__sort5'
        _VSTD::__sort5<_Compare>(__first, __first+1, __first+2, __first+3, --__last, __comp);
        ^~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__config:508:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3741:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort5(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3853:5: error: no matching function for call to '__sort3'
    __sort3<_Compare>(__first, __first+1, __j, __comp);
    ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, float>::element, gko::detail::IteratorFactory<int, float>::element> &, _ForwardIterator
      = gko::detail::IteratorFactory<int, float>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:721:71: error: invalid operands to binary expression ('const
      gko::detail::IteratorFactory<int, double>::element' and 'const gko::detail::IteratorFactory<int, double>::element')
    bool operator()(const _T1& __x, const _T1& __y) const {return __x < __y;}
                                                                  ~~~ ^ ~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3931:17: note: in instantiation of member function
      'std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>::operator()'
      requested here
            if (__comp(*--__last, *__first))
                ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, double>::element,
      gko::detail::IteratorFactory<int, double>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/utility:594:1: note: candidate template ignored: could not match
      'pair<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator< (const pair<_T1,_T2>& __x, const pair<_T1,_T2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:702:1: note: candidate template ignored: could not match
      'reverse_iterator<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const reverse_iterator<_Iter1>& __x, const reverse_iterator<_Iter2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1143:1: note: candidate template ignored: could not match
      'move_iterator<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const move_iterator<_Iter1>& __x, const move_iterator<_Iter2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1515:1: note: candidate template ignored: could not match
      '__wrap_iter<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const __wrap_iter<_Iter1>& __x, const __wrap_iter<_Iter2>& __y) _NOEXCEPT_DEBUG
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/tuple:1182:1: note: candidate template ignored: could not match
      'tuple<type-parameter-0-0...>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const tuple<_Tp...>& __x, const tuple<_Up...>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2970:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator< (const unique_ptr<_T1, _D1>& __x, const unique_ptr<_T2, _D2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3028:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const unique_ptr<_T1, _D1>& __x, nullptr_t)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3037:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(nullptr_t, const unique_ptr<_T1, _D1>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4808:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const shared_ptr<_Tp>& __x, const shared_ptr<_Up>& __y) _NOEXCEPT
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4878:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(const shared_ptr<_Tp>& __x, nullptr_t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4886:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, double>::element'
operator<(nullptr_t, const shared_ptr<_Tp>& __x) _NOEXCEPT
^
In file included from /Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:33:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/core/matrix/csr_kernels.hpp:37:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/include/ginkgo/core/base/types.hpp:43:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/complex:247:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/sstream:174:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ostream:138:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:500:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:176:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3718:20: error: no matching function for call to '__sort3'
    unsigned __r = __sort3<_Compare>(__x1, __x2, __x3, __c);
                   ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3938:20: note: in instantiation of function template specialization
      'std::__1::__sort4<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
            _VSTD::__sort4<_Compare>(__first, __first+1, __first+2, --__last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, double>::element,
      gko::detail::IteratorFactory<int, double>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3744:20: error: no matching function for call to '__sort4'
    unsigned __r = __sort4<_Compare>(__x1, __x2, __x3, __x4, __c);
                   ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3941:20: note: in instantiation of function template specialization
      'std::__1::__sort5<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
            _VSTD::__sort5<_Compare>(__first, __first+1, __first+2, __first+3, --__last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, double>::element,
      gko::detail::IteratorFactory<int, double>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3715:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort4(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3809:5: error: no matching function for call to '__sort3'
    __sort3<_Compare>(__first, __first+1, __j, __comp);
    ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3946:20: note: in instantiation of function template specialization
      'std::__1::__insertion_sort_3<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int,
      double>::element> &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
            _VSTD::__insertion_sort_3<_Compare>(__first, __last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, double>::element,
      gko::detail::IteratorFactory<int, double>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3842:9: error: no matching function for call to '__sort3'
        _VSTD::__sort3<_Compare>(__first, __first+1, --__last, __comp);
        ^~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__config:508:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4072:32: note: in instantiation of function template specialization
      'std::__1::__insertion_sort_incomplete<std::__1::__less<gko::detail::IteratorFactory<int, double>::element,
      gko::detail::IteratorFactory<int, double>::element> &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
            bool __fs = _VSTD::__insertion_sort_incomplete<_Compare>(__first, __i, __comp);
                               ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element>
      &, gko::detail::IteratorFactory<int, double>::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator, std::__1::__less<gko::detail::IteratorFactory<int, double>::element,
      gko::detail::IteratorFactory<int, double>::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, double>::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3845:9: error: no matching function for call to '__sort4'
        _VSTD::__sort4<_Compare>(__first, __first+1, __first+2, --__last, __comp);
        ^~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__config:508:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3715:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort4(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3848:9: error: no matching function for call to '__sort5'
        _VSTD::__sort5<_Compare>(__first, __first+1, __first+2, __first+3, --__last, __comp);
        ^~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/__config:508:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3741:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort5(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3853:5: error: no matching function for call to '__sort3'
    __sort3<_Compare>(__first, __first+1, __j, __comp);
    ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, double>::element, gko::detail::IteratorFactory<int, double>::element> &,
      _ForwardIterator = gko::detail::IteratorFactory<int, double>::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:721:71: error: invalid operands to binary expression ('const
      gko::detail::IteratorFactory<int, std::__1::complex<float> >::element' and 'const gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element')
    bool operator()(const _T1& __x, const _T1& __y) const {return __x < __y;}
                                                                  ~~~ ^ ~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3931:17: note: in instantiation of member function
      'std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element, gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element>::operator()' requested here
            if (__comp(*--__last, *__first))
                ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element, gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element> &, gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator, std::__1::__less<gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element, gko::detail::IteratorFactory<int, std::__1::complex<float> >::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/utility:594:1: note: candidate template ignored: could not match
      'pair<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator< (const pair<_T1,_T2>& __x, const pair<_T1,_T2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:702:1: note: candidate template ignored: could not match
      'reverse_iterator<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const reverse_iterator<_Iter1>& __x, const reverse_iterator<_Iter2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1143:1: note: candidate template ignored: could not match
      'move_iterator<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const move_iterator<_Iter1>& __x, const move_iterator<_Iter2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/iterator:1515:1: note: candidate template ignored: could not match
      '__wrap_iter<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const __wrap_iter<_Iter1>& __x, const __wrap_iter<_Iter2>& __y) _NOEXCEPT_DEBUG
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/tuple:1182:1: note: candidate template ignored: could not match
      'tuple<type-parameter-0-0...>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const tuple<_Tp...>& __x, const tuple<_Up...>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:2970:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator< (const unique_ptr<_T1, _D1>& __x, const unique_ptr<_T2, _D2>& __y)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3028:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const unique_ptr<_T1, _D1>& __x, nullptr_t)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:3037:1: note: candidate template ignored: could not match
      'unique_ptr<type-parameter-0-0, type-parameter-0-1>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(nullptr_t, const unique_ptr<_T1, _D1>& __x)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4808:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const shared_ptr<_Tp>& __x, const shared_ptr<_Up>& __y) _NOEXCEPT
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4878:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(const shared_ptr<_Tp>& __x, nullptr_t) _NOEXCEPT
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/memory:4886:1: note: candidate template ignored: could not match
      'shared_ptr<type-parameter-0-0>' against 'const gko::detail::IteratorFactory<int, std::__1::complex<float> >::element'
operator<(nullptr_t, const shared_ptr<_Tp>& __x) _NOEXCEPT
^
In file included from /Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:33:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/core/matrix/csr_kernels.hpp:37:
In file included from /Users/natalie/Library/Ginkgo/ginkgo/include/ginkgo/core/base/types.hpp:43:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/complex:247:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/sstream:174:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ostream:138:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:500:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:176:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:56:
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3718:20: error: no matching function for call to '__sort3'
    unsigned __r = __sort3<_Compare>(__x1, __x2, __x3, __c);
                   ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3938:20: note: in instantiation of function template specialization
      'std::__1::__sort4<std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element,
      gko::detail::IteratorFactory<int, std::__1::complex<float> >::element> &, gko::detail::IteratorFactory<int, std::__1::complex<float>
      >::Iterator>' requested here
            _VSTD::__sort4<_Compare>(__first, __first+1, __first+2, --__last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element, gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element> &, gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator, std::__1::__less<gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element, gko::detail::IteratorFactory<int, std::__1::complex<float> >::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3678:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element, gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element> &, _ForwardIterator = gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator]
__sort3(_ForwardIterator __x, _ForwardIterator __y, _ForwardIterator __z, _Compare __c)
^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3744:20: error: no matching function for call to '__sort4'
    unsigned __r = __sort4<_Compare>(__x1, __x2, __x3, __x4, __c);
                   ^~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3941:20: note: in instantiation of function template specialization
      'std::__1::__sort5<std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element,
      gko::detail::IteratorFactory<int, std::__1::complex<float> >::element> &, gko::detail::IteratorFactory<int, std::__1::complex<float>
      >::Iterator>' requested here
            _VSTD::__sort5<_Compare>(__first, __first+1, __first+2, __first+3, --__last, __comp);
                   ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4117:5: note: in instantiation of function template specialization
      'std::__1::__sort<std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element, gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element> &, gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator>' requested here
    __sort<_Comp_ref>(__first, __last, __comp);
    ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:4126:12: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator, std::__1::__less<gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element, gko::detail::IteratorFactory<int, std::__1::complex<float> >::element> >' requested here
    _VSTD::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
           ^
/Users/natalie/Library/Ginkgo/ginkgo/reference/matrix/csr_kernels.cpp:486:14: note: in instantiation of function template specialization
      'std::__1::sort<gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator>' requested here
        std::sort(helper.begin(), helper.end());
             ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:3715:1: note: candidate template ignored: substitution failure [with _Compare =
      std::__1::__less<gko::detail::IteratorFactory<int, std::__1::complex<float> >::element, gko::detail::IteratorFactory<int,
      std::__1::complex<float> >::element> &, _ForwardIterator = gko::detail::IteratorFactory<int, std::__1::complex<float> >::Iterator]
__sort4(_ForwardIterator __x1, _ForwardIterator __x2, _ForwardIterator __x3,
^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [reference/CMakeFiles/ginkgo_reference.dir/matrix/csr_kernels.cpp.o] Error 1
make[1]: *** [reference/CMakeFiles/ginkgo_reference.dir/all] Error 2
make: *** [all] Error 2
</details>

--- a/core/base/iterator_factory.hpp
+++ b/core/base/iterator_factory.hpp
@@ -86,6 +86,11 @@ private:
     struct element {
         ToSortType dominant;
         SecondaryType secondary;
+
+        friend bool operator<(const element &left, const element &right)
+        {
+            return left.dominant < right.dominant;
+        }
     };
 
     /**

--- a/core/test/base/iterator_factory.cpp
+++ b/core/test/base/iterator_factory.cpp
@@ -79,7 +79,8 @@ protected:
 
             // Test all combinations of the `<` operator
             if (*(begin + 1) < *begin || next_ref < curr_ref ||
-                next_ref < curr_val || next_val < curr_ref) {
+                next_ref < curr_val || next_val < curr_ref ||
+                next_val < curr_val) {
                 return false;
             }
         }

